Expand new CoreIPC serialization format to cover more of WebCoreArgumentCoders

gavin-apple opened the pull request:
#### 629347e3f7c53d8792152bbb9a08033e3a9a0eba
<pre>
Expand new CoreIPC serialization format to cover more of WebCoreArgumentCoders
<a href="https://bugs.webkit.org/show_bug.cgi?id=248397">https://bugs.webkit.org/show_bug.cgi?id=248397</a>
rdar://102717379

Reviewed by Alex Christensen.

Expand CoreIPC serialization to cover more of WebCoreArgumentCoders.
This includes:
 - VelocityData
 - MimeClassInfo
 - AuthenticationChallenge
 - DragData
 - DictationAlternative
 - UserStyleSheet
 - ScrollElasticity
 - ScrollbarMode
 - ScrollableAreaParameters
 - AbsolutePositionConstraints
 - NotificationResources

* Source/WebCore/Modules/notifications/NotificationResources.h:
(WebCore::NotificationResources::NotificationResources):
(WebCore::NotificationResources::create):
* Source/WebCore/page/scrolling/ScrollingConstraints.cpp:
(WebCore::AbsolutePositionConstraints::AbsolutePositionConstraints):
* Source/WebCore/page/scrolling/ScrollingConstraints.h:
* Source/WebCore/platform/DragData.h:
(WebCore::DragData::DragData):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/cocoa/DragDataCocoa.mm:
(WebCore::DragData::DragData):
* Source/WebCore/platform/network/soup/AuthenticationChallenge.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;VelocityData&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;VelocityData&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;MimeClassInfo&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;MimeClassInfo&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;AuthenticationChallenge&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;AuthenticationChallenge&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;DragData&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;DragData&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;DictationAlternative&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;DictationAlternative&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;UserStyleSheet&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;UserStyleSheet&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;ScrollableAreaParameters&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;ScrollableAreaParameters&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;AbsolutePositionConstraints&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;AbsolutePositionConstraints&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::NotificationResources&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::NotificationResources&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/257452@main">https://commits.webkit.org/257452@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2cdd171ba5546f78bedbfa29ea7606d2a6b504cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98940 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8149 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108352 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168607 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8702 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91471 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106321 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104640 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33619 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88411 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76468 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2054 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23035 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1960 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5132 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6920 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42509 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/notifications/NotificationResources.h:
```diff
@@ -32,10 +32,22 @@ namespace WebCore {
 
 class NotificationResources : public ThreadSafeRefCounted<NotificationResources> {
 public:
+    NotificationResources(RefPtr<Image>&& image)
```
Constructors of refcounted types should be private.

gavin-apple replied:
Done

achristensen07 commented on Source/WebCore/platform/DragData.h:
```diff
@@ -79,7 +79,14 @@ class DragData {
 
     // clientPosition is taken to be the position of the drag event within the target window, with (0,0) at the top left
     WEBCORE_EXPORT DragData(DragDataRef, const IntPoint& clientPosition, const IntPoint& globalPosition, OptionSet<DragOperation>, OptionSet<DragApplicationFlags> = { }, OptionSet<DragDestinationAction> = anyDragDestinationAction(), std::optional<PageIdentifier> pageID = std::nullopt);
+#if !PLATFORM(COCOA)
```
This looks pretty messy.  I don't think we should be introducing platform-specific constructors in this PR.

achristensen07 replied:
Instead we should just have platform-specific parameters of a shared constructor.

gavin-apple replied:
Done

achristensen07 commented on Source/WebCore/page/scrolling/ScrollingConstraints.cpp:
```diff
@@ -100,6 +100,12 @@ FloatPoint StickyPositionViewportConstraints::layerPositionForConstrainingRect(c
 
+AbsolutePositionConstraints::AbsolutePositionConstraints(const FloatSize& alignmentOffset, const FloatPoint& layerPositionAtLastLayout)
+{
+    m_alignmentOffset = alignmentOffset;
```
These should be initializers instead of in the body of the constructor.

achristensen07 commented on Source/WebCore/platform/network/soup/AuthenticationChallenge.h:
```diff
@@ -46,6 +46,12 @@ class AuthenticationChallenge final : public AuthenticationChallengeBase {
+    AuthenticationChallenge(const ProtectionSpace& protectionSpace, const Credential& proposedCredential, unsigned previousFailureCount, const ResourceResponse& response, const ResourceError& error, uint32_t tlsPasswordFlags)
+        : AuthenticationChallengeBase(protectionSpace, proposedCredential, previousFailureCount, response, error)
+    {
+        m_tlsPasswordFlags = tlsPasswordFlags; 
```
ditto

achristensen07 commented on Source/WebCore/platform/cocoa/DragDataCocoa.mm:
```diff
@@ -159,6 +159,19 @@ static inline String tiffPasteboardType()
+DragData::DragData(const String& dragStorageName, const IntPoint& clientPosition, const IntPoint& globalPosition, const Vector<String>& fileNames, OptionSet<DragOperation> sourceOperationMask, OptionSet<DragApplicationFlags> flags, OptionSet<DragDestinationAction> destinationActionMask, std::optional<PageIdentifier> pageID)
+    : m_clientPosition(clientPosition)
+    , m_globalPosition(globalPosition)
+    , m_platformDragData(0)
```
Can this be nullptr?
Can this be an initializer in the header instead of here?

gavin-apple replied:
I copied this constructor from the instance directly above which initialised m_platformDragData in the same way.

Do you mean, can we switch the declaration of m_platformDragData to default initialise to NULL and then exclude it from the initialiser list?

DragDataRef m_platformDragData { NULL };

gavin-apple committed:
Expand new CoreIPC serialization format to cover more of WebCoreArgumentCoders
https://bugs.webkit.org/show_bug.cgi?id=248397
rdar://102717379

Reviewed by Alex Christensen.

Expand CoreIPC serialization to cover more of WebCoreArgumentCoders.
This includes:
 - VelocityData
 - MimeClassInfo
 - AuthenticationChallenge
 - DragData
 - DictationAlternative
 - UserStyleSheet
 - ScrollElasticity
 - ScrollbarMode
 - ScrollableAreaParameters
 - AbsolutePositionConstraints
 - NotificationResources

* Source/WebCore/Modules/notifications/NotificationResources.h:
(WebCore::NotificationResources::NotificationResources):
(WebCore::NotificationResources::create):
* Source/WebCore/page/scrolling/ScrollingConstraints.cpp:
(WebCore::AbsolutePositionConstraints::AbsolutePositionConstraints):
* Source/WebCore/page/scrolling/ScrollingConstraints.h:
* Source/WebCore/platform/DragData.h:
(WebCore::DragData::DragData):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/cocoa/DragDataCocoa.mm:
(WebCore::DragData::DragData):
* Source/WebCore/platform/network/soup/AuthenticationChallenge.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<VelocityData>::encode): Deleted.
(IPC::ArgumentCoder<VelocityData>::decode): Deleted.
(IPC::ArgumentCoder<MimeClassInfo>::encode): Deleted.
(IPC::ArgumentCoder<MimeClassInfo>::decode): Deleted.
(IPC::ArgumentCoder<AuthenticationChallenge>::encode): Deleted.
(IPC::ArgumentCoder<AuthenticationChallenge>::decode): Deleted.
(IPC::ArgumentCoder<DragData>::encode): Deleted.
(IPC::ArgumentCoder<DragData>::decode): Deleted.
(IPC::ArgumentCoder<DictationAlternative>::encode): Deleted.
(IPC::ArgumentCoder<DictationAlternative>::decode): Deleted.
(IPC::ArgumentCoder<UserStyleSheet>::encode): Deleted.
(IPC::ArgumentCoder<UserStyleSheet>::decode): Deleted.
(IPC::ArgumentCoder<ScrollableAreaParameters>::encode): Deleted.
(IPC::ArgumentCoder<ScrollableAreaParameters>::decode): Deleted.
(IPC::ArgumentCoder<AbsolutePositionConstraints>::encode): Deleted.
(IPC::ArgumentCoder<AbsolutePositionConstraints>::decode): Deleted.
(IPC::ArgumentCoder<WebCore::NotificationResources>::encode): Deleted.
(IPC::ArgumentCoder<WebCore::NotificationResources>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/257452@main
